### ISSUE TITLE
--label inconsistent between build and run

### ISSUE THREAD

**Issue by yosifkit** [docker/docker#26027]
TLDR: `docker build --label` sends its string to be interpolated (after bash processes it) and so it  replaces bash style variables (ie `$foo`) with their value from the env/build-args, while labels passed via `docker run` are not interpreted.

**Output of `docker version`:**

```
# also tested in 1.12.1
$ docker version
Client:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.7
 Git commit:   b9f10c9
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.7
 Git commit:   b9f10c9
 Built:        
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
$ docker info
Containers: 7
 Running: 7
 Paused: 0
 Stopped: 0
Images: 6651
Server Version: 1.11.2
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 4.6.5-gentoo
Operating System: Gentoo/Linux
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 30.88 GiB
Name: isengard
ID: UOCM:3F65:5FZC:6H5L:W3HY:34G4:A5XZ:SUOV:S2D4:XQTO:4KGA:6XSE
Docker Root Dir: /mnt/spare/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Physical machine. 

**Steps to reproduce the issue:**
 1. Build a Dockerfile with `--label`

  ```console
$ docker build --label 'cli_label$foo=$bar' -t test-labels - <<'EOF'
FROM alpine
EOF
  ```

 1. Try running an image with some labels

  ```console
$ docker run --label foo='$bar' --label 'foo$two'='more $bar' --name labels -it --rm test-labels sh
/ #
  ```

 1. Check `docker inspect` output (see next section)

**Describe the results you received:**
First, look at the output of the above `docker build`.

```console
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM alpine
 ---> 45a09e37e616
Step 2 : LABEL "cli_label$foo" "$bar"
 ---> Running in 89af0220fbc7
 ---> 97b8ad7418da
Removing intermediate container 89af0220fbc7
Successfully built 97b8ad7418da
```

Ok, lets check the image and container with inspect:

```console
$ # image
$ docker inspect --format "{{ json .Config.Labels }}" test-labels | jq
{
  "cli_label": ""
}
$ # container
$ docker inspect --format "{{ json .Config.Labels }}" labels | jq
{
  "cli_label": "",
  "foo": "$bar",
  "foo$two": "more $bar"
}
```

**Describe the results you expected:**

I would expect build to not interpret variables in strings just like `docker run`. So the build above would've resulted in a config like this:
```json
{
  "cli_label$foo": "$bar"
}
```

I know it is possible to do via Dockerfile with with single quotes: `LABEL 'cli_label$foo'='$bar'`.

This is an inconsistent interface for `--label` between `build` and `run`.

**Comment by justincormack** [comment]
Yes, I think that is a bug, although it will induce a behaviour change which it is possible people have been relying on, which is a bit difficult, but I think it is probably better to fix it.

I think that there should be a JSON form of label, ie `LABEL ["key=value", "key2=value2"]` and that the `--label` option on `docker build` should use that. The JSON forms do not run through the shell so there is no shell expansion then, which would solve this, and also provide an option to not have expansion in the Dockerfile without having to use single quotes.
